csrc/python_frontend/segmentation.h

```diff
@@ -0,0 +1,221 @@
+//! corresponds with [T0, T2, T3] in the original fusion. Therefore, the segment
+//! index to original index map has two entries.
+//!
+//! Segment Index : Original Index Mapping
+//! --------------------------------------
+//! T0 : T0 --- The first tensor argument for the original fusion.
+//! T2 : T3 --- The broadcasted, reduction tensor is this segment's output.
+//!
+//! The second segment is the pointwise addition with the broadcasted reduction.
+//! It corresponds with [T1, T3, T4] in the original fusion.
+//!
+//! Segment Index : Original Index Mapping
+//! --------------------------------------
+//! T0 : T1 --- The second tensor argument for the original fusion.
+//! T1 : T3 --- The broadcasted, reduction tensor, which is the output from the
+//!             first segment.
+//! T2 : T4 --- The pointwise addition, which is the output for the original
+//!             fusion.
+//! ===========================================================================
+class SegmentationState {
+ public:
+  // setupSegmentation runs the segmentation algorithm on CPP Fusion to create
+  // SegmentedFusion. It returns the number of segments in SegmentedFusion.
+  //
+  // Details:
+  //  1) Clone preschedFusion CPP Fusion.
+  //  2) Concretize fusion using input arguments.
+  //  3) Given the map_value_to_original_fid, the IRCloner returned by
+  //     Fusion::copy, AND symbolic_to_concrete map returned by concretization
+  //     pass, create a mapping from cloned Vals to original fusion state
+  //     indices
+  //  4) Get extents for cloned fusion
+  //  5) Create SchedulerRuntimeInfo
+  //  6) Run segmentation algorithm using cloned fusion, input arguments, and
+  //  scheduler runtime infomation.
+  //  7) Get sequential order of fusion segments using prepareGroupOrder.
+  //  8) Return the number of segments created by segmentation algorithm.
+  int64_t setupSegmentation(
+      Fusion* fusion,
+      const std::unordered_map<const Val*, int64_t>& map_value_to_original_fid,
+      const at::ArrayRef<c10::IValue>& inputs);
+
+  // prepareGroupOrder is similar to prepareRuntimeOrder. It generates the
+  // topological order of SegmentedGroups in SegmentedFusion.
+  //
+  // Details:
+  //  1) Gather initial inputs for SegmentedFusion.
+  //  2) Gather IterDomain extents from the tensor input arguments.
+  //  3) Track the run status of all SegmentedGroups in SegmentedFusion
+  //  4) While not all the SegmentedGroups are run:
+  //  5)   For each SegmentedGroup:
+  //  6)     Skip SegmentedGroup if it is already run
+  //  7)     Skip SegmentedGroup if inputs are not ready
+  //  8)     Add SegmentedGroup to group_run_order_. Mark all outputs of
+  //         SegmentedGroup as ready.
+  //  9)   End For
+  //  10)  Fail if none of the SegmentedGroups are available to run.
+  //  11) End While
+  void prepareGroupOrder();
+
+ private:
+  // Clone of original fusion for segmentation
+  std::unique_ptr<Fusion> segment_fusion_ = nullptr;
```

Comment:
Would `cloned_fusion_` be a better name here, this is confusing since we also have `segmented_fusion_`?